### DISCUSSION TITLE
pico read write serial

### DISCUSSION THREAD

**Discussion by sepogramy** [micropython/micropython#9161] [General]

**Comment by jimmo** [comment]
The rp2040 (i.e. Pico) has some constraints on which pins can be used for a given peripheral (e.g. UART).

The rules for the rp2040 are:

```c
#define IS_VALID_PERIPH(uart, pin)  (((((pin) + 4) & 8) >> 3) == (uart))
#define IS_VALID_TX(uart, pin)      (((pin) & 3) == 0 && IS_VALID_PERIPH(uart, pin))
#define IS_VALID_RX(uart, pin)      (((pin) & 3) == 1 && IS_VALID_PERIPH(uart, pin))
#define IS_VALID_CTS(uart, pin)     (((pin) & 3) == 2 && IS_VALID_PERIPH(uart, pin))
#define IS_VALID_RTS(uart, pin)     (((pin) & 3) == 3 && IS_VALID_PERIPH(uart, pin))
```

This translates to:
 - tx must be 0, 4, 8, 12...etc for uart 0, 1, 2, 3...etc
 - rx must be 1, 5, 9, 13...etc for uart 0, 1, 2, 3...etc

So in your case, you'll need to use `tx=Pin(4)` and `rx=Pin(5)` if you want to use `UART(1)`

Also please use triple-backticks for formatting code (I've updated your post). Single backticks are used for `print("inline snippets")` whereas triple is used for blocks.

**Reply by sepogramy** [reply]
,

**Comment by robert-hh** [comment]
About this sequence:
```py
print(uart.write(bytes(data)))
print(uart.read())
```
The python code returns immediately after the uart.write() statemen and **before** all data has been written. So the uart.read() statement returns what has been received up to the, which is None. You have to wait after uart.write() until all data has been sent and the answer arrived. Doing that, you can insert a time.sleep_ms(), or add a timeout when creating the UART object, or use uart.any() to poll for incoming data.
Edit: I see that you have already a timeout set. Are you sure that you use the proper pins? The pin numbers for Pin(x) are GPIO numbers, not board pin numbers.

**Reply by sepogramy** [reply]
.

**Reply by robert-hh** [reply]
How long do you wait after the write() for data coming in? Sending 8 bytes at 9600 baud takes ate least 8 ms.
What kind of hardware is connected? Does it use 'TTL' levels at 3.3V, or V24 levels?

**Comment by sepogramy** [comment]
.

**Reply by sepogramy** [reply]
.

**Reply by robert-hh** [reply]
The max232 should not get hot. If you connect rx and tx at the RS232 side, you  should receive back what you have sent.

**Reply by sepogramy** [reply]
.

**Reply by robert-hh** [reply]
Did you try the loopback tests? 
1. The first step is loopback at the pico board. Disconnect the Max232 and connect GPIO0 and GPIO1. Then you should be able to read back what you have sent. Make sure you get it.
2. Then connec the MAX232, connect rx and tx at the V24 side and run the loopback test again. If that fails, try to swap the connections to GPIO0 and GPIO1. If the loopback test still fails, then there is a problem with the MAX232.

**Reply by scruss** [reply]
@sepogramy - please stop editing your posts to delete the contents. The discussions are here to help other people too, and if you delete your content you're wasting the time of the volunteers who answer here

**Reply by Lobo-T** [reply]
What's the point of this behaviour?
One of the features of git is that you can't actually delete anything anyway:
![image](https://user-images.githubusercontent.com/13045607/187993364-d1d01c34-b4ef-4678-a2fd-3edb7360269a.png)

**Reply by jimmo** [reply]
@sepogramy Yes please stop deleting your posts. As @Lobo-T points out the posts aren't actually deleted because they're still available in the edit history.
And like @scruss says, the point of these forums is to make the answers available to anyone who has the same problem in the future. The people in this thread have put time and effort into answering your questions.

If you do need to delete your post for some reason then the thread can be deleted (although it will still be in the email history), but again we'd prefer not to have to do that.

**Reply by sepogramy** [reply]
tried to delete to create new post to make this more clear

**Comment by sepogramy** [comment]
the hardware works with all loopback tests, the code in python works
`import serial
  def serial_communication():
  ser = serial.Serial()
  ser.port = "COM1"
  ser.baudrate = 9600
  ser.bytesize = serial.EIGHTBITS 
  ser.parity = serial.PARITY_NONE 
  ser.stopbits = serial.STOPBITS_ONE 
  ser.timeout = 2          
  ser.xonxoff = False    
  ser.rtscts = False     
  ser.dsrdtr = False       
  ser.writeTimeout = 2    

  try:

      ser.open()

  except Exception as e:

      print ("error open serial port: " + str(e))

      exit()

  if ser.isOpen():
      ser.flushInput() 
      ser.flushOutput()
      ser.write(serial.to_bytes([0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00]))   
      response=ser.read(200) 
      print(response.hex())`

the code in micropython doesnt work and return b'00'
dont understand the difference 
`
  from machine import UART, Pin
  import ubinascii
  import time
  import machine
  uart=UART(0, baudrate=9600, bits=8, parity=None, stop=1, tx=Pin(0), rx=Pin(1), txbuf=256, rxbuf=256, timeout=2000)
  data=[0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00]
  print(uart.write(bytes(data)))
  print(respond_binary=uart.read())
  respond_hex=ubinascii.hexlify(bytes(respond_binary, 'utf-8'))
  print(respond_hex)
    
`
